runtime-sdk/modules/contracts/src/abi/oasis/crypto.rs

```diff
@@ -91,33 +92,67 @@ impl<Cfg: Config> OasisV1<Cfg> {
+                        .map_err(|_| wasm3::Trap::Abort)?
+                        .to_vec()
+                        .into();
+                    if context.0 != 0 && context.1 != 0 && matches!(kind, SignatureKind::Sr25519) {
+                        let context = Region::from_arg(context)
                             .as_slice(&memory)
                             .map_err(|_| wasm3::Trap::Abort)?;
-                        let signature: signature::Signature = Region::from_arg(signature)
-                            .as_slice(&memory)
-                            .map_err(|_| wasm3::Trap::Abort)?
-                            .to_vec()
-                            .into();
-                        if context.0 != 0
-                            && context.1 != 0
-                            && matches!(kind, SignatureKind::Sr25519)
-                        {
-                            let context = Region::from_arg(context)
-                                .as_slice(&memory)
-                                .map_err(|_| wasm3::Trap::Abort)?;
-                            Ok(1 - key.verify(context, message, &signature).is_ok() as u32)
-                        } else {
-                            Ok(1 - key.verify_raw(message, &signature).is_ok() as u32)
-                        }
-                    })?
+                        Ok(1 - key.verify(context, message, &signature).is_ok() as u32)
+                    } else {
+                        Ok(1 - key.verify_raw(message, &signature).is_ok() as u32)
+                    }
+                })?
+            },
+        );
+
+        // crypto.random_bytes(dst) -> bytes_written
+        let _ = instance.link_function(
+            "crypto",
+            "random_bytes",
+            |ctx,
+             ((pers_ptr, pers_len, dst_ptr, dst_len),): ((u32, u32, u32, u32),)|
+             -> Result<u32, wasm3::Trap> {
+                // Make sure function was called in valid context.
+                let ec = ctx.context.ok_or(wasm3::Trap::Abort)?;
+
+                let num_bytes = dst_len.min(1024 /* 1 KiB */);
+
+                // Charge gas.
+                let cost = ec
+                    .params
+                    .gas_costs
+                    .wasm_crypto_random_bytes_byte
+                    .checked_mul(num_bytes as u64 + pers_len as u64)
+                    .and_then(|g| g.checked_add(ec.params.gas_costs.wasm_crypto_random_bytes_base))
+                    .unwrap_or(u64::max_value()); // This will certainly exhaust the gas limit.
+                gas::use_gas(ctx.instance, cost)?;
+
+                let rt = ctx.instance.runtime();
+                rt.try_with_memory(|mut memory| -> Result<_, wasm3::Trap> {
+                    let pers = Region::from_arg((pers_ptr, pers_len))
+                        .as_slice(&memory)
+                        .map_err(|_| wasm3::Trap::Abort)?;
+                    let mut rng = ec.tx_context.rng(pers).map_err(|_| wasm3::Trap::Abort)?;
```

Comment:
Actually, this also needs to propagate the error, similarly to how it is done in `storage::get_instance_store`, e.g.:
https://github.com/oasisprotocol/oasis-sdk/blob/04944cbb7a3cf346acb2693b0a7dfda4ab0f9f63/runtime-sdk/modules/contracts/src/abi/oasis/storage.rs#L186-L194

Reply:
Ok, now it batch aborts any time a RNG/KM error happens.